pilot/pkg/networking/core/v1alpha3/tracing.go

```diff
@@ -0,0 +1,480 @@
+		hcm.Tracing = &hpb.HttpConnectionManager_Tracing{}
+		// TODO: transition to configuring providers from proxy config here?
+		// something like: configureFromProxyConfig(tracingCfg, opts.proxy.Metadata.ProxyConfig.Tracing)
+	}
+
+	// gracefully fallback to MeshConfig configuration. It will act as an implicit
+	// parent configuration during transition period.
+	configureSampling(hcm.Tracing, tracingCfg.RandomSamplingPercentage, proxyCfg)
+	configureCustomTags(hcm.Tracing, tracingCfg.CustomTags, proxyCfg)
+
+	// if there ia configured max tag length somewhere, fallback to it.
+	if hcm.GetTracing().GetMaxPathTagLength() == nil && proxyCfg.GetTracing().GetMaxPathTagLength() != 0 {
+		hcm.Tracing.MaxPathTagLength = wrapperspb.UInt32(proxyCfg.GetTracing().MaxPathTagLength)
+	}
+}
+
+// TODO: follow-on work to enable bootstrapping of clusters for $(HOST_IP):PORT addresses.
+
+func configureFromProviderConfig(pushCtx *model.PushContext, meta map[string]interface{},
+	providerCfg *meshconfig.MeshConfig_ExtensionProvider) (*hpb.HttpConnectionManager_Tracing, error) {
+	switch provider := providerCfg.Provider.(type) {
+	case *meshconfig.MeshConfig_ExtensionProvider_Zipkin:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Zipkin.Service, provider.Zipkin.Port, provider.Zipkin.MaxTagLength, zipkinConfigGen)
+	case *meshconfig.MeshConfig_ExtensionProvider_Datadog:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Datadog.Service, provider.Datadog.Port, provider.Datadog.MaxTagLength, datadogConfigGen)
+	case *meshconfig.MeshConfig_ExtensionProvider_Lightstep:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Lightstep.Service, provider.Lightstep.Port, provider.Lightstep.MaxTagLength,
+			func(clusterName string) (*anypb.Any, error) {
+				lc := &tracingcfg.LightstepConfig{
+					CollectorCluster: clusterName,
+					AccessTokenFile:  provider.Lightstep.AccessToken,
+				}
+				return anypb.New(lc)
+			})
+
+	case *meshconfig.MeshConfig_ExtensionProvider_Opencensus:
+		return buildHCMTracingOpenCensus(providerCfg.Name, provider.Opencensus.MaxTagLength, func() (*anypb.Any, error) {
+			oc := &tracingcfg.OpenCensusConfig{
+				OcagentAddress:         fmt.Sprintf("%s:%d", provider.Opencensus.Service, provider.Opencensus.Port),
+				OcagentExporterEnabled: true,
+				IncomingTraceContext:   convert(provider.Opencensus.Context),
+				OutgoingTraceContext:   convert(provider.Opencensus.Context),
+			}
+
+			return anypb.New(oc)
+		})
+
+	case *meshconfig.MeshConfig_ExtensionProvider_Stackdriver:
+		return buildHCMTracingOpenCensus(providerCfg.Name, provider.Stackdriver.MaxTagLength, func() (*anypb.Any, error) {
+			proj, ok := meta[platform.GCPProject].(string)
+			if !ok {
+				proj, ok = meta[platform.GCPProjectNumber].(string)
+			}
+			if !ok {
+				return nil, fmt.Errorf("could not configure Stackdriver tracer - unknown project id")
+			}
+			sd := &tracingcfg.OpenCensusConfig{
```

Comment:
I think we need to configure the gRPC service credential based on sts port enablement within the proxy node metadata.

Reply:
ok. i'll look into that now. i'm not super familiar with that logic.

Reply:
added sts support. please take a look. i copied from gcp_bootstrap. hopefully it is correct.